### DISCUSSION TITLE
Routing computation performance: Can it be faster?

### DISCUSSION THREAD

**Discussion by brad-4k** [valhalla/valhalla#3743] [Q&A]
Hello, and thank you for all of your hard work on Valhalla.

We have setup a server and are running it through its paces and have found that the time to compute a route, as compared to other systems, can be "long".   I'm wondering if there are some settings or tweaks that I can do to decrease the time it takes to compute a route or is it something fundamental?

I've included a graph of our results with a lot of queries of various length and the amount of time it takes to return the results.

Thank you, again.

![Screen Shot 2022-09-11 at 8 21 05 PM](https://user-images.githubusercontent.com/78495562/189567962-bdb05271-1c99-4743-a544-f7dcd32dfc39.png)

**Comment by nilsnolde** [comment]
What’s your machine specs? Are users requesting concurrently or it’s only one request after the other? 

One way, regardless of the answers, to increase performance: don’t do docker! Don’t even do valhalla_service. Set up all workers with their own processes and have prime_server proxy between them. It’s a bit more complex but when you look at loki_worker.cc etc, it might become clearer.

**Comment by nilsnolde** [comment]
Also, are you using HTTP I assume? Is the graph showing the timings from the client or the server? Was it benchmarked with every request creating a new connection?

**Reply by brad-4k** [reply]
Thank you Nils for the response.

I'll try to answer all your questions here.

Yes, I'm using docker on an AWS ECS system.  Rules are setup to allow it to expand but ...

In the tests above, it's one query at a time.  Client side, using HTTP and timing the total request time.  A new request each time.  Both systems I tested are hosted on instances with the same spec's.

It all came to a head when we tried it with an actual feed, 4,000,000 requests over an hour or so and latency started to become greater than 5 seconds (my timelimit with my frontend).  So the timeouts climbed geometrically.

WRT your suggested solution, that is really above my knowledge base right now with Valhalla.  Is there some kind of example for setting something similar up?

Thank you

**Reply by nilsnolde** [reply]
4 Mio per hour?! That’s more than 1000/sec. Is Google Maps switching to Valhalla?😄 you’re load testing I assume?

Even if you have 32 threads (sorry, never worked with AWS, and hopefully never will, so I have no idea what ECS exactly means), you’d look at 30 requests per thread per second, ie 30 ms for each request, which is too ambitious. 64 threads might work if the route is not too long/complex. Even then just the round tripping over http might get ya.

IMO the best way here is to scale horizontally, add more servers to the pool, a load balancer in front and a native setup if possible (can always ask your favorite consultants;)). Eg IIRC mapbox uses dozens of Valhalla servers (for more than just routing though).

**Reply by brad-4k** [reply]
Nils, that 4mil may be over a couple hours.  I'm still trying to dig into that, but yes, our system will be a firehose of queries.  A very good throughput and performance case for Valhalla.

Our target response time is to be no greater than 500ms per query, and scaling is an option.  I'm no AWS expert either, but we have those and are starting to grab their attention.

I can't make any $ decisions but tell me more about consulting (I'm assuming you mean with respect to Valhalla).  Load balancing is in there as well as a pool and a set of rules for expanding the server base in real-time.

That said, you still have me intrigued with your first response on a reconfiguration of the "services".  Are there examples (or just an example) of what you described?

**Reply by nilsnolde** [reply]
> tell me more about consulting

You can contact us on enquiry@gis-ops.com if you're interested.

> Are there examples (or just an example) of what you described?

Not that I'm aware of, sorry.

**Reply by brad-4k** [reply]
Nils,

So, if we could simplify things.  At this time, we are only using the total distance and duration of the route.  Not the legs.  Can I get Valhalla to just return the total Summary? The idea being that it might save some time not generating the turn-by-turn, etc.

**Reply by nilsnolde** [reply]
You can try the matrix endpoint (doesn’t make much sense for 1:1 though) and/or no narrative for the route endpoint (you’d still get the geometry, can’t avoid that).

**Comment by ImreSamu** [comment]
My tips:
* Adding a proxy + multiple Valhalla servers
* Maybe you can tune the default valhalla config ( cache,  search, .. ) for your use case
    *  ( see the config generator: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config )
* Rebuild the binaries with `-O3 -march=x86-64-v3` + LTO 
* ...

**Reply by brad-4k** [reply]
Thank you Imre for your answer.  Although I've built Valhalla in our initial testing we are now relying on a prebuilt docker image, which I currently don't know how it was optimized.

**Comment by brad-4k** [comment]
Just a follow up.  Looks like we are hitting our preliminary numbers.  This was not 4mil / hr but more like 1000 / min.  We reconfigured our AWS setup in a much "smarter" way and added a few more instances.
Thank you all, again, for your positive feedback.

**Reply by abhishekkushwaha-fareye** [reply]
Would love to hear more about this, I am facing the same challenges while trying to scale things up.

**Reply by kinkard** [reply]
@abhishekkushwaha-fareye In AWS-native setup just define ALB and target group with Valhalla instances and all over sudden you have easy to scale Valhalla. With tileset on S3 and proper script that downloads it on an instance startup it is even possible to setup automatic scaling up and down with ALB + target groups.

**Reply by abhishekkushwaha-fareye** [reply]
Thank you, I'll give it a try.
